changefeedccl: fix alter changefeed to handle targets with fully qualified names

sherman-grewal opened the pull request:
changefeedccl: fix alter changefeed to handle targets with
fully qualified names

Resolves #79502

Currently, the ALTER CHANGEFEED statement would not work
with changefeeds that use fully qualified names in their
CREATE CHANGEFEED statement. This is because when we try
to include the existing targets, we would drop the
database/schema name that the user provided. Hence, when
we perform validation checks on the existing + new
changefeed targets, we would not be able to resolve the
existing targets that used fully qualified names and we
would return an error.

In this PR we ensure that we add each existing target
with their fully qualified name so that we can resolve
these targets in our validation checks. A consequence of
this change is that now every changefeed will display
the fully qualified name of every target in the
SHOW CHANGEFEED JOB query.

Release note (enterprise change): Fix the issue of not
being able to resolve existing targets when performing
validation checks by adding the fully qualified name
to the changefeed statement. As a result, every altered
changefeed will display the fully qualified name of every
target in the SHOW CHANGEFEED JOB query.

Release justification: Low risk bug fix.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79542)
<!-- Reviewable:end -->

miretskiy commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -317,8 +317,18 @@ func generateNewTargets(
+		if err != nil {
+			return nil, nil, hlc.Timestamp{}, err
+
+		}
```
I can see why we are doing what we're doing here... But the conversion from current TargetSpecification -> ChangefeedTarget looses StatementTimeName I think.

Consider what happens when you created changefeed on table "foo", but then, later renamed it to "bar".  Wouldn't you loose "foo" statement time name?  Even ignoring the fact that in the meantime you could have created another foo table.  

I think that's the problem @HonoreDB mentioned yesterday.

Ignoring for a second whether we should allow such operation in the first place, I think we need to deal with this rename problem for now.

I think we can probably do something about this.  Let's add a "resolved flavor" to ChangefeedTarget -- somewhat similar to "resolve name" and "unresolved name".

In `pkg/sql/sem/tree/changefeed.go` let's add an un-exported field to `ChangefeedTarget` containing resolved jobspb.ChangefeedTargetSpecification... We almost certainly won't be able to add this field directly -- i.e. using jobspb inside tree package is probably not a good idea... but still, we can probably do:

```
type ChangefeedTarget struct {
	TableName  TablePattern
	FamilyName Name
  resolvedTarget interface{}
}

// MakeResolvedChangefeedTarget construct changefeed target
// that has already been resolved.
func MakeResolvedChangefeedTarget(resolved interface) ChangefeedTarget {
   return ChangefeedTarget{resolvedTarget: resolved}
}

func (ct *ChangefeedTarget) ResolvedSpec() interface {
  return ct.resolved
}
```

I don't think we need anything more in the tree package.  In particular, you don't need
to worry about being able to format that "resolved" field as a string because this interface is pretty much added for internal communication between alter statement and the rest of the changefeed code.  

Alter changefeed, instead of doing the above code, would construct MakeResolvedChangefeedTarget using existing "target".
What else needs to change? Probably just the `getTargetsAndTables` in changefeed_statement so that it looks to see if target has non nil ResolvedSpec and just uses it.

What do you think?
@HonoreDB does that sound reasonable to you?

HonoreDB replied:
That makes sense to me, yeah--we're telling changefeed_stmt to reuse an old name resolution instead of re-doing it. If it feels too weird to mess with tree, an equivalent thing could be to build a `originalSpecs` map of tree.ChangefeedTarget to ChangefeedTargetSpecification and pass it around.

miretskiy replied:
Yeah, if there is a way to avoid that -- great... It's just we still need to construct ChangefeedStatement to pass to the rest of the code.  I think it's important to use the "original" logic in changefeed_statement wrt to validation, job record creation, etc.  So, one way or another, this information needs to me in the form of a ChangefeedStatement.
Now, in the scheduled backup (and elsewhere), I used a trick where I introduced annotated statement:
```
// annotatedBackupStatement is a tree.Backup, optionally
// annotated with the scheduling information.
type annotatedBackupStatement struct {
	*tree.Backup
	*jobs.CreatedByInfo
}
```
And back would see if the tree.Backup is an annotated statement (i.e. the call is made from the scheduled jobs system), and do something else.  I suppose we can choose to maintain original targets by passing e.g.
```
type alterChangefeedStmt struct {
   *tree.ChangefeedStatement
   originalSpecs ...
}
```
And do something appropriate with this.

HonoreDB commented on pkg/ccl/changefeedccl/alter_changefeed_test.go:
```diff
@@ -609,6 +609,47 @@ func TestAlterChangefeedAddTargetErrors(t *testing.T) {
 	t.Run(`kafka`, kafkaTest(testFn, feedTestNoTenants))
 }
 
+func TestAlterChangefeedDatabaseQualifiedNames(t *testing.T) {
```
As you mess with this, probably a good idea to have test cases like these:

```
CREATE CHANGEFEED FOR movr.drivers, printr.drivers
USE movr; CREATE CHANGEFEED FOR movr.drivers FAMILY f1, drivers FAMILY f2 // refers to the same table twice in two different ways
```

No need to add test cases where we haven't defined the right behavior yet, but

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79542)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @stevendanna)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 329 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79542#-N--ktnG8fs45rc2ES8R-r1-329:-N-5KfYqA04p-y-OmBoB:b-hojp9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/12e29a494f5148b47861a6ca1e30548a61a0bd2d/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L329)):*
I ended up going down the path of creating that `originalSpecs` mapping between a changefeed target and its corresponding target specification, and then inserting that map into an annotated changefeed statement. Thanks @miretskiy @honoredb for the suggestions!

> Consider what happens when you created changefeed on table "foo", but then, later renamed it to "bar". Wouldn't you loose "foo" statement time name? Even ignoring the fact that in the meantime you could have created another foo table.

I added a test for the scenario you mentioned above.
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79542#-N--ktnG8fs45rc2ES8S-r1-612:-N-5LA7L2SiVBkfX4ZBd:bgjv86) ([raw file](https://github.com/cockroachdb/cockroach/blob/12e29a494f5148b47861a6ca1e30548a61a0bd2d/pkg/ccl/changefeedccl/alter_changefeed_test.go#L612)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

As you mess with this, probably a good idea to have test cases like these:

```
CREATE CHANGEFEED FOR movr.drivers, printr.drivers
USE movr; CREATE CHANGEFEED FOR movr.drivers FAMILY f1, drivers FAMILY f2 // refers to the same table twice in two different ways
```

No need to add test cases where we haven't defined the right behavior yet, but for stuff where there's a clear answer...
</blockquote></details>

Great suggestion! I added test cases for the scenarios you called out (and more!)


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -36,6 +36,22 @@ func init() {
+	default:
+		return nil
+	}
+}
```
Let's move this code to changefeed_stmt.go.

miretskiy commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -308,38 +342,53 @@ func generateNewTargets(
+		k := targetKey{TableID: targetSpec.TableID, FamilyName: tree.Name(targetSpec.FamilyName)}
+		desc := descResolver.DescByID[targetSpec.TableID].(catalog.TableDescriptor)
+
+		tbName, err := getQualifiedTableNameObj(ctx, p.ExecCfg(), p.ExtendedEvalContext().Txn, desc)
```
Why do we need to resolve anything in the prevDetails? Why can't we just use prevDetails.TargetSpecifications?

miretskiy replied:
It also appears to me that perhaps having a tableID -> target ChangefeedTargetSpecification is a better structure?

sherman-grewal replied:
> It also appears to me that perhaps having a tableID -> target ChangefeedTargetSpecification is a better structure?

Since column families would have the same tableID, but different statement time names, I'm not sure if this mapping would work.

> Why do we need to resolve anything in the prevDetails? Why can't we just use prevDetails.TargetSpecifications?

We can't just use the prevDetails.TargetSpecifications because we would have no way of knowing which target specification to use in prevDetails (if any) for any ChangefeedTarget in `rawTargets` in the function `getTargetsAndTables`. Recall that targets can also get dropped, so we can't use prevDetails.TargetSpecifications in that case.

If we wanted to go down the route of just using prevDetails.TargetSpecifications, we would also need a way to find out which targets in `rawTargets` are going to use prevDetails and which ones will not. We could set up `originalSpecs` so that it includes all of the TargetSpecifications that we would need (so `changefeed_stmt` would not even need to call `getTargetsAndTables`), but then we would be performing a lot of duplicate work that is in the function `getTargetsAndTables`.

Another nice added bonus of performing a resolve is that we can update the changefeed description to the new target name in the event of a table rename.

sherman-grewal replied:
@HonoreDB might want to speak more on these questions

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79542)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 356 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79542#-N--ktnG8fs45rc2ES8R-r2-356:-N-8a1Q31p9Fu40dKUmm:b105tta) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd18cf45a9638f0292a9164de3cb53b83bb7e2d3/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L356)):*
<details><summary><i>Previously, sherman-grewal (Sherman) wrote…</i></summary><blockquote>

@HonoreDB might want to speak more on these questions
</blockquote></details>

FWIW Sherman talked me into a lot of this yesterday (or we talked each other into different aspects of it), and I endorse this response. I didn't really appreciate before that the most important part of the job description string is that you can copy/paste it to create a new job (except for redacted secrets), which means that updating the target names is more of a feature than a hacky side effect.
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 685 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79542#-N-8f6qf7mykzqy7JtxL:-N-8f6qf7mykzqy7JtxM:bhzl0zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/967da04ff7c8fa35d34e35a51deaabe06bca66e4/pkg/ccl/changefeedccl/alter_changefeed_test.go#L685)):*
> ```Go
> 		assertPayloads(t, testFeed, []string{
> 			`drivers: [1]->{"after": {"id": 1, "name": "Alice"}, "before": null}`,
> 		})
> ```

Could you also test here that existing unqualified targets don't get re-resolved to the wrong database? Something like
```
USE movr; CREATE CHANGEFEED for drivers WITH diff;
pause
USE new_movr. ALTER CHANGEFEED %d UNSET diff;
resume
INSERT INTO movr.drivers values (1,'bob')
assertPayloads(drivers - > 1, bob)
```
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 788 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79542#-N-8cin6-_Bhwn_dDL0v:-N-8cin6-_Bhwn_dDL0w:b-7ne4oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/967da04ff7c8fa35d34e35a51deaabe06bca66e4/pkg/ccl/changefeedccl/alter_changefeed_test.go#L788)):*
> ```Go
> 		})
> 
> 		//expectResolvedTimestamp(t, testFeed)
> ```

commented code


<!-- Sent from Reviewable.io -->

miretskiy commented:
Thanks for explaining.  This :lgtm: to me, but please get a vote from @HonoreDB

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79542)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)

___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 685 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79542#-N-8f6qf7mykzqy7JtxL:-N-976WuBnceVHbDgv1T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/967da04ff7c8fa35d34e35a51deaabe06bca66e4/pkg/ccl/changefeedccl/alter_changefeed_test.go#L685)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Could you also test here that existing unqualified targets don't get re-resolved to the wrong database? Something like
```
USE movr; CREATE CHANGEFEED for drivers WITH diff;
pause
USE new_movr. ALTER CHANGEFEED %d UNSET diff;
resume
INSERT INTO movr.drivers values (1,'bob')
assertPayloads(drivers - > 1, bob)
```

</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 788 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/79542#-N-8cin6-_Bhwn_dDL0v:-N-977A87ztn3_rCVYzF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/967da04ff7c8fa35d34e35a51deaabe06bca66e4/pkg/ccl/changefeedccl/alter_changefeed_test.go#L788)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

commented code
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
changefeedccl: fix alter changefeed to handle targets with
fully qualified names

Currently, the ALTER CHANGEFEED statement would not work
with changefeeds that use fully qualified names in their
CREATE CHANGEFEED statement. This is because when we try
to include the existing targets, we would drop the
database/schema name that the user provided. Hence, when
we perform validation checks on the existing + new
changefeed targets, we would not be able to resolve the
existing targets that used fully qualified names and we
would return an error.

In this PR we ensure that we add each existing target
with their fully qualified name so that we can resolve
these targets in our validation checks. A consequence of
this change is that now every changefeed will display
the fully qualified name of every target in the
SHOW CHANGEFEED JOB query.

Release note (enterprise change): Fix the issue of not
being able to resolve existing targets when performing
validation checks by adding the fully qualified name
to the changefeed statement. As a result, every altered
changefeed will display the fully qualified name of every
target in the SHOW CHANGEFEED JOB query.

Release justification: Low risk bug fix.

HonoreDB reviewed (approved):
> Reviewed 1 of 3 files at r1, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79542)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @stevendanna)



<!-- Sent from Reviewable.io -->

sherman-grewal commented:
bors r+
